Monika-After-Story/MonikaModDev
# Consumables

# Introducing `MASConsumable`s

## Additions:
- `MASConsumable` class
- Thermos for taking consumable drinks out with you (Thanks @Professor-JMO) (These can also be spritepacked)
- Generalized logic for checking if we should have a consumable
- Generalized labels for finished preparing, having, and getting consumables
- Thermos mug gift handler

### Global Function Additions:
#### `safeDel`:
- Function to safely `del` variables from the persistent
#### `mas_getConsumable()`:
- Gets a consumable by consumable id, regardless of type
#### `mas_generateShoppingList()`:
- Generates a 'shopping list' of items Monika is running low of and exports it to the characters folder for the player

## Changes:
- Dockstat IO loop now has Monika step away from the desk as she gets ready (so she can put her drink in a thermos if applicable. Can be expanded for changing clothes come outfit rand)
- `zz_dump.rpy` now reports for all consumables
- Quetzal and chocolates reactions now change how they show the sprite when Monika is eating something
- Quetzal and chocolates will share a **ex-prop** with consumable foods (`left-desk-acs`)
- If Monika is eating something on load, quetzal will not have a chance of showing up on her desk

## Removals:
- Old consumable code
- Old hotchoc/coffee vars

## Testing:
NOTE: This WILL require unstable for more long term testing, of course.
- Verify no crashes (as usual)
- Verify that when calling the update script, if you had given coffee/hotchoc before, the respective one(s) are enabled and there is between 40-60 servings left of each
- Verify that after calling the update script, none of the old coffee/hotchoc vars still exist
- Verify that coffee still shows up within its normal hours (5 AM to 12 PM)
- Verify hotchoc still only shows up in winter in its normal hours (4 PM to 11 PM)
- Verify that loading in after 10, most of the time, coffee should already be on Monika's desk (without her having to make/get any) assuming she decided to have some
- Verify that loading in after 7, most of the time,  hotchoc should already be on Monika's desk (without her having to make/get any) assuming she decided to have some
- Verify that when using the taking you somewhere farewell with a drink out, Monika will return with a thermos (assuming it has been gifted. If it has not been gifted, she'll return with the cup)
- Verify that when Monika is out of a consumable (no servings left), she will no longer prepare/have it. You can set the `servings_left` value yourself by adjusting it in the consumable's key in the `persistent._mas_consumable_map` variable)
- Verify that Monika alerts the player via a queued topic after she finishes having a consumable (finished having label) that she's running low on some things (this only applies if she's low on one or two consumables
- Verify that Monika alerts the player via the same queued topic, but points out the specifics of what she's low on via a `shopping_list.txt` file in the `\characters` folder.
- Verify that gift reactions for coffee/hotchoc are different if Monika is out of the consumable you're gifting/already has some of the consumable you're gifting
- I will provide two more consumables via Discord for testing with. These both test two different things:

## Additional Consumable Testing:
### Tea:
- Tests multiple hour ranges for consumables

### Cherry Cheesecake
- Tests the food type consumable. Works best with tea as their times overlap so their interactions can be monitored
- Tests the max re-serve count. Monika should only re-serve herself twice for this.
